File: go/hash_test.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type TestManifest []struct {
	Name          string `json:"name"`
	File          string `json:"file"`
	FileMin       string `json:"file-min"`
	FilePretty    string `json:"file-pretty"`
	FileCanonical string `json:"file-canonical"`
	Format        string `json:"format"`
	SHA256        string `json:"sha256"`
	Source        string `json:"source"`
}

func TestHash(t *testing.T) {
	// Path to the shared vectors at the root of the repo
	manifestPath := filepath.Join("..", "test-vectors", "manifest.json")

	content, err := os.ReadFile(manifestPath)
	if err != nil {
		t.Fatalf("Failed to load manifest: %v", err)
	}

	var manifest TestManifest
	json.Unmarshal(content, &manifest)

	for _, testCase := range manifest {
		//t.Run(testCase.Name, func(t *testing.T) {
		if testCase.Format != "json" {
			t.Skip("XML not yet implemented")
		}

		files := []struct {
			name string
			path string
		}{
			{"original", testCase.File},
			{"min", testCase.FileMin},
			{"pretty", testCase.FilePretty},
			{"canonical", testCase.FileCanonical},
		}

		for _, f := range files {
			t.Run(testCase.Name+"_"+f.name, func(t *testing.T) {
				path := filepath.Join("..", "test-vectors", f.path)
				bytes, _ := os.ReadFile(path)

				computedHash, err := ComputeJSONHash(bytes)
				if err != nil {
					t.Fatalf("Hash error: %v", err)
				}

				if computedHash != testCase.SHA256 {
					t.Errorf("Hash Mismatch!\nExpected: %s\nGot: %s\nFile: %s", testCase.SHA256, computedHash, f.path)
				}
			})
		}
		//})
	}
}

File: javascript/hash.test.js
import { describe, test, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import { computeJsonHash } from './hash.js';

const manifestPath = join('..', 'test-vectors', 'manifest.json');
const manifest = JSON.parse(readFileSync(manifestPath, 'utf8'));

describe('SBOM Hashing Interoperability', () => {
  manifest.forEach((testCase) => {
    if (testCase.format !== 'json') {
      return;
    }

    describe(testCase.name, () => {
      // Map the manifest keys to a clean array for testing
      const variants = [
        { name: 'original', path: testCase.file },
        { name: 'min', path: testCase['file-min'] },
        { name: 'pretty', path: testCase['file-pretty'] },
        { name: 'canonical', path: testCase['file-canonical'] },
      ];

      test.each(variants)('Variant: $name', (variant) => {
        const fullPath = join('..', 'test-vectors', variant.path);
        const content = readFileSync(fullPath);

        const hash = computeJsonHash(content);

        expect(hash).toBe(testCase.sha256);
      });
    });
  });
});
